custom/icds/location_reassignment/utils.py

```diff
@@ -52,28 +52,31 @@ def get_household_and_child_case_ids_by_owner(domain, household_case_id, owner_i
 
 
 def get_household_child_cases_by_owner(domain, household_case_id, owner_id, case_types):
-    def get_child_cases(ids):
+    def get_child_cases(ids, exclude_ids):
         return [case for case in
                 CaseAccessorSQL.get_reverse_indexed_cases(domain, ids)
-                if case.owner_id == owner_id]
+                if case.case_id not in exclude_ids and case.owner_id == owner_id]
 
-    cases = []
+    # keep a list of case ids already iterated to avoid duplicates
+    iterated_case_ids = set()
+    cases = set()
     parent_case_ids = [household_case_id]
     while parent_case_ids:
-        child_cases = get_child_cases(parent_case_ids)
+        child_cases = get_child_cases(parent_case_ids, iterated_case_ids)
+
         if child_cases:
-            parent_case_ids = [case.case_id for case in child_cases]
-            if case_types:
-                cases.extend([case for case in child_cases if case.type in case_types])
-            else:
-                cases.extend(child_cases)
+            child_case_ids = [case.case_id for case in child_cases]
+            iterated_case_ids.update(child_case_ids)
+            cases.update([case for case in child_cases if case.type in case_types])
+            parent_case_ids = child_case_ids
```

Comment:
As an alternative to passing `iterated_case_ids` to `get_child_cases`, you could subtract them from `child_case_ids` here:
```python
parent_case_ids = child_case_ids - iterated_case_ids
```
(it would have to be defined as a set for this to work, of course, but that shouldn't be a problem)

Reply:
Oh, my approach prevents you from doing the lookup multiple times with a single ID, but it doesn't deduplicate the lookup results, as this approach does.  For that you'd have to deduplicate before updating `cases`

Reply:
`iterated_case_ids` have already been excluded so they wont be present in child_case_ids to that set subtraction wont be useful?

Reply:
They're excluded by passing it in to `get_child_cases` and skipping them there - I was saying you could instead do a simple set subtraction here, but that won't work with the current state, since `cases` is no longer a data structure which excludes duplicates.